CodeCoverageQualityGate/CodeCoverageQualityService.ts

```diff
@@ -287,9 +322,181 @@ export class CodeCoverageQualityService {
+
+        tl.debug("Evaluating quality gate");
+        if (qualityGateResult.codeCoverage == 'N/A' || parseFloat(qualityGateResult.codeCoverage) >= this.threshold) {
+            qualityGateResult.status = QualityGateStatusEnum.PASSED;
+            tl.setResult(tl.TaskResult.Succeeded, `Quality gate '${this.getQualityGateIdentification()}' passed`);
+        } else {
+            switch (this.buildStatus) {
+                case BuildStatusEnum.UNSTABLE:
+                    qualityGateResult.status = QualityGateStatusEnum.UNSTABLE;
+                    tl.setResult(tl.TaskResult.SucceededWithIssues, `Quality gate '${this.getQualityGateIdentification()}' failed: build result is UNSTABLE`);
+                    break;
+                case BuildStatusEnum.FAILED:
+                    qualityGateResult.status = QualityGateStatusEnum.FAILED;
+                    tl.setResult(tl.TaskResult.Failed, `Quality gate '${this.getQualityGateIdentification()}' failed: build result is FAILED`);
+                    break;
+                default:
+                    // User will never come here
+                    tl.error(`The build status should be unstable or failed instead of ${this.buildStatus}`);
+            }
+        }
+        tl.debug(`Quality Gate ${qualityGateResult.status} - ${this.type} code coverage: ${qualityGateResult.codeCoverage} (${coveredLines}/${coverableLines}) - Threshold: ${this.threshold}%`);
+        return qualityGateResult;
+    }
+
+    private getOverallCodeCoverage = (coberturaContentString: string): void => {
+        const saxParser = sax.parser(true);
+        saxParser.onopentag = (node) => {
+            if (node.name == 'coverage') {
+                this.overallCoveredLines = parseInt(<string>node.attributes['lines-covered']);
+                this.overallCoverabledLines = parseInt(<string>node.attributes['lines-valid']);
+            }
+        }
+
+        saxParser.onend = () => {
+            // do nothing
+        };
+
+        saxParser.onerror = (e) => {
+            tl.warning('Failed to parse cobertura code coverage report content.');
+            console.error(e);
+        };
+
+        saxParser.write(coberturaContentString).close();
+    }
+
+
     private getQualityGateIdentification() {
         const referencePipeline: string = this.originalReferencePipelineName ? ", Reference pipeline: " + this.originalReferencePipelineName : "";
         const referenceBuild: string = this.originalReferenceBuildNumber ? ", Reference build: " + this.originalReferenceBuildNumber : "";
         return "Type: " + this.type + ", Threshold: " + this.threshold + referencePipeline + referenceBuild;
     }
+
+    private getModifiedCodeCoverage = (referenceBuildCoverageReport: string, currentBuildCoverageReport: string): void => {
+        const currentBuildCodeFileContainer: CodeFile[] = [];
+        this.getCoverageDataFromReport(currentBuildCoverageReport, currentBuildCodeFileContainer);
```

Comment:
Same here, I suggest to make the return type of "getCoverageDataFromReport" to be "CodeFile[]" instead of pass it in.